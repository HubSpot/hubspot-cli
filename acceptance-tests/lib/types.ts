export interface TestConfig {
  debug: boolean;
  useInstalled: boolean;
  cliPath: string;
  personalAccessKey: string;
  portalId: string;
  qa: boolean;
  githubToken: string;
}

export interface CLI {
  execute: (
    args: string[],
    inputs?: string[],
    opts?: object
  ) => Promise<unknown>;
  executeWithTestConfig: (
    args: string[],
    inputs?: string[],
    opts?: object
  ) => Promise<unknown>;
}
