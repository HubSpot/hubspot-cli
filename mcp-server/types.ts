import {
  McpServer,
  RegisteredTool,
} from '@modelcontextprotocol/sdk/server/mcp.js';

export class Tool<InputSchema, ResponseType = TextContentResponse> {
  protected mcpServer: McpServer;
  constructor(mcpServer: McpServer) {
    this.mcpServer = mcpServer;
  }
  register(): RegisteredTool {
    throw new Error('Must implement register');
  }

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  handler(input: InputSchema): ResponseType | Promise<ResponseType> {
    throw new Error('Must implement handler');
  }
}

export type TextContent = { type: 'text'; text: string };

export type TextContentResponse = { content: TextContent[] };
