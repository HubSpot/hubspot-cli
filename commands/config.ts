// @ts-nocheck
const { i18n } = require('../lib/lang');
const set = require('./config/set');

const i18nKey = 'commands.config';

exports.command = 'config';
exports.describe = i18n(`${i18nKey}.describe`);

exports.builder = yargs => {
  yargs.command(set).demandCommand(1, '');

  return yargs;
};
