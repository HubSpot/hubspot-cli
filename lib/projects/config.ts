import fs from 'fs-extra';
import path from 'path';
import findup from 'findup-sync';
import { logger } from '@hubspot/local-dev-lib/logger';
import { getAbsoluteFilePath, getCwd } from '@hubspot/local-dev-lib/path';

import { ProjectConfig } from '../../types/Projects';
import { PROJECT_CONFIG_FILE } from '../constants';
import { lib } from '../../lang/en';
import { uiCommandReference } from '../ui';
import { EXIT_CODES } from '../enums/exitCodes';

export function writeProjectConfig(
  configPath: string,
  config: ProjectConfig
): boolean {
  try {
    fs.ensureFileSync(configPath);
    fs.writeFileSync(configPath, JSON.stringify(config, null, 2));
    logger.debug(`Wrote project config at ${configPath}`);
  } catch (e) {
    logger.debug(e);
    return false;
  }
  return true;
}

export function getIsInProject(dir?: string): boolean {
  const configPath = getProjectConfigPath(dir);
  return !!configPath;
}

function getProjectConfigPath(dir?: string): string | null {
  const projectDir = dir ? getAbsoluteFilePath(dir) : getCwd();

  const configPath = findup(PROJECT_CONFIG_FILE, {
    cwd: projectDir,
    nocase: true,
  });

  return configPath;
}

export interface LoadedProjectConfig {
  projectDir: string | null;
  projectConfig: ProjectConfig | null;
}

export async function getProjectConfig(
  dir?: string
): Promise<LoadedProjectConfig> {
  const configPath = getProjectConfigPath(dir);
  if (!configPath) {
    return { projectConfig: null, projectDir: null };
  }

  try {
    const config = fs.readFileSync(configPath);
    const projectConfig: ProjectConfig = JSON.parse(config.toString());
    return {
      projectDir: path.dirname(configPath),
      projectConfig,
    };
  } catch (e) {
    logger.error('Could not read from project config');
    return { projectConfig: null, projectDir: null };
  }
}

export function validateProjectConfig(
  projectConfig: ProjectConfig | null,
  projectDir: string | null
): asserts projectConfig is ProjectConfig {
  if (!projectConfig || !projectDir) {
    logger.error(
      lib.projects.validateProjectConfig.configNotFound(
        uiCommandReference('hs project create')
      )
    );
    return process.exit(EXIT_CODES.ERROR);
  }

  if (!projectConfig.name || !projectConfig.srcDir) {
    logger.error(lib.projects.validateProjectConfig.configMissingFields);
    return process.exit(EXIT_CODES.ERROR);
  }

  const resolvedPath = path.resolve(projectDir, projectConfig.srcDir);
  if (!resolvedPath.startsWith(projectDir)) {
    const projectConfigFile = path.relative(
      '.',
      path.join(projectDir, PROJECT_CONFIG_FILE)
    );
    logger.error(
      lib.projects.validateProjectConfig.srcOutsideProjectDir(
        projectConfigFile,
        projectConfig.srcDir
      )
    );
    return process.exit(EXIT_CODES.ERROR);
  }

  if (!fs.existsSync(resolvedPath)) {
    logger.error(
      lib.projects.validateProjectConfig.srcDirNotFound(
        projectConfig.srcDir,
        projectDir
      )
    );

    return process.exit(EXIT_CODES.ERROR);
  }
}
