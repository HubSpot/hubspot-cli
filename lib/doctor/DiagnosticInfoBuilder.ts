import { getProjectConfig } from '../projects/config.js';
import { fetchProject } from '@hubspot/local-dev-lib/api/projects';
import path from 'path';
import { pkg } from '../jsonLoader.js';
import { uiLogger } from '../ui/logger.js';
import { Environment } from '@hubspot/local-dev-lib/types/Config';
import {
  AccessToken,
  AccountType,
  AuthType,
} from '@hubspot/local-dev-lib/types/Accounts';
import { Project } from '@hubspot/local-dev-lib/types/Project';
import {
  getAccountId,
  getDefaultAccountOverrideFilePath,
  isConfigFlagEnabled,
} from '@hubspot/local-dev-lib/config';
import { getAccountConfig, getConfigPath } from '@hubspot/local-dev-lib/config';
import { getAccessToken } from '@hubspot/local-dev-lib/personalAccessKey';
import { walk } from '@hubspot/local-dev-lib/fs';
import util from 'util';
import { exec as execAsync } from 'node:child_process';
import process from 'process';
import {
  CMS_ASSETS_FILE,
  LEGACY_PRIVATE_APP_FILE,
  LEGACY_PUBLIC_APP_FILE,
  LEGACY_SERVERLESS_FILE,
  PROJECT_CONFIG_FILE,
  THEME_FILE,
} from '../constants.js';

export type ProjectConfig = Awaited<ReturnType<typeof getProjectConfig>>;

// This needs to be hardcoded since we are using it in the TS type
const hubspotCli = '@hubspot/cli';

interface FilesInfo {
  files: string[];
  configFiles: string[];
  packageFiles: string[];
  packageLockFiles: string[];
  envFiles: string[];
  jsonFiles: string[];
}

export interface DiagnosticInfo extends FilesInfo {
  path?: string;
  versions: { [hubspotCli]: string; node: string; npm: string | null };
  config: string | null;
  defaultAccountOverrideFile: string | null | undefined;
  configSettings: { [key: string]: unknown };
  project: {
    details?: Project;
    config?: ProjectConfig;
  };
  arch: typeof process.arch;
  platform: typeof process.platform;
  account: {
    name?: string;
    accountId?: number | null;
    scopeGroups?: AccessToken['scopeGroups'];
    enabledFeatures?: AccessToken['enabledFeatures'];
    accountType?: AccessToken['accountType'];
    authType?: AuthType;
  };
  diagnosis?: string;
  errorCount?: number;
  warningCount?: number;
}

const configFiles = [
  LEGACY_SERVERLESS_FILE,
  PROJECT_CONFIG_FILE,
  LEGACY_PRIVATE_APP_FILE,
  LEGACY_PUBLIC_APP_FILE,
  THEME_FILE,
  CMS_ASSETS_FILE,
];

export class DiagnosticInfoBuilder {
  accountId: number | null;
  readonly configSettings: { [key: string]: unknown };
  readonly env?: Environment;
  readonly authType?: AuthType;
  readonly accountType?: AccountType;
  readonly personalAccessKey?: string;
  private _projectConfig?: ProjectConfig;
  private accessToken?: AccessToken;
  private projectDetails?: Project;
  private files?: string[];
  readonly processInfo: NodeJS.Process;

  constructor(processInfo: NodeJS.Process) {
    this.accountId = getAccountId();
    const accountConfig = getAccountConfig(this.accountId!);
    this.configSettings = {
      httpUseLocalhost: isConfigFlagEnabled('httpUseLocalhost'),
    };
    this.env = accountConfig?.env;
    this.authType = accountConfig?.authType;
    this.accountType = accountConfig?.accountType;
    this.personalAccessKey = accountConfig?.personalAccessKey;
    this.processInfo = processInfo;
  }

  async generateDiagnosticInfo(): Promise<DiagnosticInfo> {
    this._projectConfig = await getProjectConfig();

    if (this._projectConfig?.projectConfig) {
      await this.fetchProjectDetails();
      await this.fetchAccessToken();
    }

    if (this._projectConfig?.projectDir) {
      await this.fetchProjectFilenames();
    }

    const {
      platform,
      arch,
      versions: { node },
      mainModule,
    } = this.processInfo;

    return {
      platform,
      arch,
      path: mainModule?.path,
      config: getConfigPath(),
      defaultAccountOverrideFile: getDefaultAccountOverrideFilePath(),
      configSettings: this.configSettings,
      versions: {
        [hubspotCli]: pkg.version,
        node,
        npm: await this.getNpmVersion(),
      },
      account: {
        accountId: this.accountId,
        accountType: this.accountType,
        authType: this.authType,
        name: this.accessToken?.hubName,
        scopeGroups: this.accessToken?.scopeGroups,
        enabledFeatures: this.accessToken?.enabledFeatures,
      },
      project: {
        config: this._projectConfig,
        details: this.projectDetails,
      },
      ...this.generateFilesArrays(),
      files: this.files || [],
    };
  }

  private async fetchProjectDetails(): Promise<void> {
    try {
      const { data } = await fetchProject(
        this.accountId!,
        // We check that config exists before running this function
        this._projectConfig!.projectConfig!.name
      );
      this.projectDetails = data;
    } catch (e) {
      uiLogger.debug(e);
    }
  }

  async fetchAccessToken(): Promise<AccessToken | undefined> {
    try {
      this.accessToken = await getAccessToken(
        this.personalAccessKey!,
        this.env,
        this.accountId!
      );
    } catch (e) {
      uiLogger.debug(e);
    }

    return this.accessToken;
  }

  private async fetchProjectFilenames(): Promise<void> {
    try {
      // We check that projectDir exists before running this function
      this.files = (await walk(this._projectConfig!.projectDir!))
        .filter(file => !path.dirname(file).includes('node_modules'))
        .map(filename =>
          path.relative(this._projectConfig!.projectDir!, filename)
        );
    } catch (e) {
      uiLogger.debug(e);
    }
  }

  private async getNpmVersion(): Promise<string | null> {
    const exec = util.promisify(execAsync);
    try {
      const { stdout } = await exec('npm --version');
      return stdout.toString().trim();
    } catch (e) {
      uiLogger.debug(e);
      return null;
    }
  }

  private generateFilesArrays(): FilesInfo {
    const output: FilesInfo = {
      files: this.files || [],
      configFiles: [],
      packageFiles: [],
      packageLockFiles: [],
      envFiles: [],
      jsonFiles: [],
    };

    if (!this.files) {
      return output;
    }

    return this.files.reduce((acc: FilesInfo, file) => {
      const { base } = path.parse(file);

      if (base === 'package.json') {
        acc.packageFiles.push(file);
      } else if (configFiles.includes(base)) {
        acc.configFiles.push(file);
        if (file.endsWith('.json')) {
          acc.jsonFiles.push(file);
        }
      } else if (base === 'package-lock.json') {
        acc.packageLockFiles.push(file);
      } else if (file.endsWith('.env')) {
        acc.envFiles.push(file);
      } else if (file.endsWith('.json')) {
        acc.jsonFiles.push(file);
      }

      return acc;
    }, output);
  }
}
