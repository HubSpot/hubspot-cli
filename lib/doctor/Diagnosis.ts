import { prefixOptions } from '../ui/spinniesUtils';
import { bold, green, red } from 'chalk';
import { helpers } from '../interpolation';
import { DiagnosticInfo } from './DiagnosticInfoBuilder';
import { uiAccountDescription } from '../ui';
const { i18n } = require('../lang');

interface DiagnosisOptions {
  diagnosticInfo: DiagnosticInfo;
  accountId: number | null;
}

interface Section {
  type: 'error' | 'warning' | 'success';
  message: string;
  secondaryMessaging?: string;
}

type prefixes = {
  [prefix in Section['type']]: string;
};

interface DiagnosisCategory {
  header: string;
  subheaders?: string[];
  sections: Section[];
}

interface DiagnosisCategories {
  cli: DiagnosisCategory;
  project: DiagnosisCategory;
  cliConfig: DiagnosisCategory;
}

const i18nKey = `lib.doctor.diagnosis`;

export class Diagnosis {
  private readonly prefixes: prefixes;
  private readonly diagnosis: DiagnosisCategories;
  private readonly indentation = '  ';
  private errorCount = 0;
  private warningCount = 0;

  constructor({ diagnosticInfo, accountId }: DiagnosisOptions) {
    const { succeedPrefix, failPrefix } = prefixOptions({});

    this.prefixes = {
      success: green(succeedPrefix),
      error: red(failPrefix),
      warning: helpers.orange('!'),
    };

    this.diagnosis = {
      cli: {
        header: i18n(`${i18nKey}.cli.header`),
        sections: [],
      },
      cliConfig: {
        header: i18n(`${i18nKey}.cliConfig.header`),
        sections: [],
      },
      project: {
        header: i18n(`${i18nKey}.projectConfig.header`),
        subheaders: [
          i18n(`${i18nKey}.projectConfig.projectDirSubHeader`, {
            projectDir: diagnosticInfo.project?.config?.projectDir,
          }),
          i18n(`${i18nKey}.projectConfig.projectNameSubHeader`, {
            projectName: diagnosticInfo.project?.config?.projectConfig?.name,
          }),
        ],
        sections: [],
      },
    };

    if (diagnosticInfo.config) {
      this.diagnosis.cliConfig.subheaders = [
        i18n(`${i18nKey}.cliConfig.configFileSubHeader`, {
          filename: diagnosticInfo.config,
        }),
        i18n(`${i18nKey}.cliConfig.defaultAccountSubHeader`, {
          accountDetails: uiAccountDescription(accountId!),
        }),
      ];
    }
  }

  private indent(level: number): string {
    return this.indentation.repeat(level);
  }

  getErrorCount(): number {
    return this.errorCount;
  }

  getWarningCount(): number {
    return this.warningCount;
  }

  addCliSection(section: Section): void {
    this.diagnosis.cli.sections.push(section);
  }

  addProjectSection(section: Section): void {
    this.diagnosis.project.sections.push(section);
  }

  addCLIConfigSection(section: Section): void {
    this.diagnosis.cliConfig.sections.push(section);
  }

  toString(): string {
    const output = [];
    for (const value of Object.values(this.diagnosis)) {
      const section = this.generateSections(value);
      if (section) {
        output.push(section);
      }
    }

    if (output.length === 0) {
      return '';
    }

    output.push('');
    output.push(
      i18n(`${i18nKey}.counts.errors`, {
        count: this.errorCount,
      })
    );
    output.push(
      i18n(`${i18nKey}.counts.warnings`, {
        count: this.warningCount,
      })
    );
    output.push('');

    return output.join('\n');
  }

  private generateSections(category: DiagnosisCategory): string {
    const output = [];

    if (category.sections && category.sections.length === 0) {
      return '';
    }

    output.push(`\n${bold(category.header)}`);

    (category.subheaders || []).forEach(subheader => {
      output.push(`${subheader}`);
    });

    category.sections.forEach(section => {
      if (section.type === 'error') {
        this.errorCount++;
      } else if (section.type === 'warning') {
        this.warningCount++;
      }

      output.push(
        `${this.indent(1)}${this.prefixes[section.type]} ${section.message}`
      );
      if (section.secondaryMessaging) {
        output.push(`${this.indent(2)}- ${section.secondaryMessaging}`);
      }
    });

    return output.join('\n');
  }
}
