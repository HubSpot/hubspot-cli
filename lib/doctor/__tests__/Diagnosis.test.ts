import { CLIAccount } from '@hubspot/local-dev-lib/types/Accounts';

jest.mock('@hubspot/local-dev-lib/config');

import { Diagnosis } from '../Diagnosis';
import { DiagnosticInfo } from '../DiagnosticInfoBuilder';
import { getAccountConfig as __getAccountConfig } from '@hubspot/local-dev-lib/config';
import stripAnsi from 'strip-ansi';

const getAccountConfig = __getAccountConfig as jest.MockedFunction<
  typeof __getAccountConfig
>;

describe('lib/doctor/Diagnosis', () => {
  const diagnosticInfo: DiagnosticInfo = {
    account: {},
    arch: process.arch,
    config: 'path/to/config.json',
    configFiles: [],
    envFiles: [],
    files: [],
    jsonFiles: [],
    packageFiles: [],
    packageLockFiles: [],
    platform: process.platform,
    project: {
      config: {
        projectDir: 'project-dir',
        projectConfig: {
          name: 'Super cool project',
        },
      },
    },
    versions: { '@hubspot/cli': '123', node: '18.0.1', npm: '8.0.0' },
  };

  const accountId = 123456;

  beforeEach(() => {
    getAccountConfig.mockReturnValue({
      accountType: 'STANDARD',
      name: 'Standard Account',
    } as CLIAccount);
  });

  describe('toString', () => {
    it('should return an empty diagnosis when no sections have been added', () => {
      const diagnosis = new Diagnosis({ diagnosticInfo, accountId });
      const output = diagnosis.toString();
      expect(output).toEqual('');
    });

    it('should generate the diagnosis output', () => {
      const diagnosis = new Diagnosis({ diagnosticInfo, accountId });

      const cliMessage = 'Important CLI Message';
      const cliSecondaryMessage = 'The CLI Section is Showing';

      diagnosis.addCliSection({
        type: 'success',
        message: cliMessage,
        secondaryMessaging: cliSecondaryMessage,
      });

      const cliConfigMessage = 'Important CLI Config Message';
      const cliConfigSecondaryMessage = 'The CLI Config Section is Showing';

      diagnosis.addCLIConfigSection({
        type: 'error',
        message: cliConfigMessage,
        secondaryMessaging: cliConfigSecondaryMessage,
      });

      const projectMessage = 'Important Project Message';
      const projectSecondaryMessage = 'The Project Section is Showing';

      diagnosis.addProjectSection({
        type: 'warning',
        message: projectMessage,
        secondaryMessaging: projectSecondaryMessage,
      });

      expect(stripAnsi(diagnosis.toString())).toMatchSnapshot();
    });

    it('should generate categories that have sections', () => {
      const diagnosis = new Diagnosis({ diagnosticInfo, accountId });

      const cliMessage = 'Important CLI Message';
      const cliSecondaryMessage = 'The CLI Section is Showing';

      diagnosis.addCliSection({
        type: 'success',
        message: cliMessage,
        secondaryMessaging: cliSecondaryMessage,
      });

      expect(stripAnsi(diagnosis.toString())).toMatchSnapshot();
    });
  });
});
