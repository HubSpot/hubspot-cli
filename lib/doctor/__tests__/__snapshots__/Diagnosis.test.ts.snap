// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lib/doctor/Diagnosis toString should generate categories that have sections 1`] = `
"
HubSpot CLI install
  ✓ Important CLI Message
    - The CLI Section is Showing

Errors: 0
Warning: 0
"
`;

exports[`lib/doctor/Diagnosis toString should generate the diagnosis output 1`] = `
"
HubSpot CLI install
  ✓ Important CLI Message
    - The CLI Section is Showing

CLI configuration
Config File: path/to/config.json
Default Account: Standard Account [standard] (123456)
  ✖ Important CLI Config Message
    - The CLI Config Section is Showing

Project configuration
Project dir: project-dir
Project name: Super cool project
  ! Important Project Message
    - The Project Section is Showing

Errors: 1
Warning: 1
"
`;
